### PR TITLE
cleanup(generator): better library name parsing for nested services

### PR DESCRIPTION
**Pull request by dbolduc**
Part of the work for #10170 

I would rather assume that the `product_path` matches `google/cloud/service($|/**)` than assume that it matches `**/service`

(I am sure my regex syntax is wrong, but you get the point)

- - -
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10171)

### CODE REVIEW

File: generator/internal/scaffold_generator.cc [hunk 1]
```diff
@@ -36,9 +36,10 @@ auto constexpr kApiIndexFilename = "api-index-v1.json";
 
 std::string LibraryName(
     google::cloud::cpp::generator::ServiceConfiguration const& service) {
-  auto const l = service.product_path().find_last_of('/');
-  if (l == std::string::npos) return {};
-  return service.product_path().substr(l + 1);
+  std::vector<std::string> v = absl::StrSplit(service.product_path(), '/');
+  // Assume we are given: google/cloud/<library>/...
```
Comment (devbww) [inline generator/internal/scaffold_generator.cc:40]: Why assume, when you can check?

File: generator/internal/scaffold_generator.cc [hunk 2]
```diff
@@ -36,9 +36,10 @@ auto constexpr kApiIndexFilename = "api-index-v1.json";
...
+  if (v.size() < 2) return {};
```
Comment (devbww) [inline generator/internal/scaffold_generator.cc:41]: `<= 2`, or `< 3` I presume?

File: generator/internal/scaffold_generator.cc [hunk 3]
```diff
@@ -36,9 +36,10 @@ auto constexpr kApiIndexFilename = "api-index-v1.json";
 
 std::string LibraryName(
     google::cloud::cpp::generator::ServiceConfiguration const& service) {
-  auto const l = service.product_path().find_last_of('/');
-  if (l == std::string::npos) return {};
-  return service.product_path().substr(l + 1);
+  std::vector<std::string> v = absl::StrSplit(service.product_path(), '/');
```
Comment (devbww) [inline generator/internal/scaffold_generator.cc:39]: The PR description mentions "regex syntax".  A remnant perhaps?

File: generator/internal/scaffold_generator.cc (hunk 2 above)
Comment (dbolduc) [reply generator/internal/scaffold_generator.cc:41]: Lol......🐒

File: generator/internal/scaffold_generator.cc (hunk 1 above)
Comment (dbolduc) [reply generator/internal/scaffold_generator.cc]: Fixed.

File: generator/internal/scaffold_generator.cc (hunk 2 above)
Comment (dbolduc) [reply generator/internal/scaffold_generator.cc]: Fixed

**Review by scotthart — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10171)* status: 0 of 5 files reviewed, 4 unresolved discussions (waiting on @dbolduc and @devbww)

*[a discussion](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10171#-NFyKFP_EO_AgC8NMv1U:-NFyKFP_EO_AgC8NMv1V:b-6dexzo) (no related file):*
Might want to add some tests for LibraryName as well.

___

**Comment by dbolduc** [comment]
> Might want to add some tests for LibraryName as well

Done.

File: generator/internal/scaffold_generator.cc (hunk 3 above)
Comment (devbww) [reply generator/internal/scaffold_generator.cc:39]: Ahh!  Silly me.

**Review by devbww — APPROVED** [review]

**Review by scotthart — APPROVED** [review]

**Pull request merged**